akka-actor-tests/src/test/scala/akka/pattern/RetrySpec.scala

```diff
@@ -98,6 +106,64 @@ class RetrySpec extends AkkaSpec with RetrySupport {
         intercept[IllegalStateException] { Await.result(retried, remaining) }.getMessage should ===("6")
       }
     }
+
+    "return the result after retrying with a backoff" in {
+      @volatile var failCount = 0
+
+      def attempt() = {
+        failCount += 1
+        Future.failed(new IllegalStateException(failCount.toString))
+      }
+
+      retry(
+        attempt,
+        3,
+        100 milliseconds,
+        2
+      )
+
+      within(100 milliseconds, 3 second) {
+        eventually {
+          failCount should ===(2)
+        }
+      }
+
+      within(200 milliseconds, 3 second) {
+        eventually {
+          failCount should ===(3)
+        }
+      }
+
+      within(400 milliseconds, 3 second) {
+        eventually {
+          failCount should ===(4)
+        }
+      }
+    }
+
+    "execute the provided function during every retry" in {
+      val s = new ListBuffer[String]()
+      val attempts = 3
+      val delay = 100 milliseconds
+      val backoff = 2
+
+      retryF(
+        () ⇒ Future.failed(new Exception("oups")),
+        attempts,
+        delay,
+        backoff
+      ) {
+          case (t, currentRetry) ⇒
+            s.append(s"${t.getMessage}, will retry in ${delay * math.pow(backoff, attempts - currentRetry + 1)}")
```

Comment:
is the callback invoke when the exception occurs or before next attempt (after delay)?

Reply:
Good point. I will change it after we decide how the function should look